File: Python/pandas/09_hourly_distribution.py
"""
09_hourly_distribution.py
-------------------------
At what hours of the day do members and casual riders most frequently
start rides?

Key finding: Members show a dual peak at ~8 AM and ~5 PM — a classic
commute pattern. Casual riders show a single broad afternoon peak
around 3–5 PM — consistent with leisure use.

pandas notes:
  - .dt.hour extracts the integer hour (0–23) from a datetime column.
    Equivalent to SQL's HOUR() / DATEPART(HOUR, ...).
  - groupby().agg() with a dict of named aggregations produces the hourly
    counts for both rider types in one pass.
  - Percentage of type: divide each hour's member/casual count by the
    total for that type — makes the two curves directly comparable
    regardless of the overall member/casual volume difference.
  - ASCII bar chart added to make the commute vs leisure pattern visible
    directly in the terminal, matching the DuckDB module output.

Expected output (matching SQL Server, DuckDB, and R):
  Member peaks at hours 8 and 17 (8 AM and 5 PM) — commute pattern
  Casual peak at hours 15–17 (3–5 PM) — leisure pattern
"""

import pandas as pd
from utils import load_trips

trips = load_trips()

# ---------------------------------------------------------------------------
# Extract hour and count by rider type
# ---------------------------------------------------------------------------

trips["hour_of_day"] = trips["started_at"].dt.hour

hourly = (
    trips.groupby(["hour_of_day", "member_casual"])
    .size()
    .unstack(fill_value=0)
)
hourly["total"] = hourly.sum(axis=1)

# Percentage of each rider type's total (makes curves comparable)
total_member = hourly["member"].sum()
total_casual = hourly["casual"].sum()

hourly["member_pct"] = hourly["member"] / total_member * 100
hourly["casual_pct"] = hourly["casual"] / total_casual * 100

# ---------------------------------------------------------------------------
# ASCII bar chart output
# ---------------------------------------------------------------------------

BAR_SCALE = 0.2  # 1 character per 0.2%

print("=" * 72)
print("  Hourly Ride Distribution  (% of each group's total rides)")
print("=" * 72)
print(f"  {'Hr':>5}  {'Member%':>7}  {'Casual%':>7}  {'Member bar':<25} {'Casual bar'}")
print(f"  {'-'*70}")

for hour, row in hourly.iterrows():
    mem_pct = row["member_pct"]
    cas_pct = row["casual_pct"]
    mem_bar = "█" * int(mem_pct / BAR_SCALE)
    cas_bar = "█" * int(cas_pct / BAR_SCALE)
    label   = f"{hour:02d}:00"
    print(f"  {label:>5}  {mem_pct:>7.2f}%  {cas_pct:>7.2f}%  {mem_bar:<25} {cas_bar}")

print("=" * 72)
print("  Members:  dual peaks at 08:00 and 17:00 → commute pattern")
print("  Casuals:  single broad peak at 15:00–17:00 → leisure pattern")
print("=" * 72)

File: R/09_hourly_distribution.R
# 09_hourly_distribution.R
# Cyclistic Bike-Share Case Study — R Module
#
# Ride volume by hour of day, split by rider type.
# Key finding: members show dual peaks at 8 AM and 5 PM (commute pattern).
#              Casuals show a single afternoon peak (leisure pattern).
#
# Introduces: lubridate::hour(), geom_area(), faceted area chart

library(tidyverse)

csv_folder <- file.path(dirname(dirname(rstudioapi::getSourceEditorContext()$path)),
                        "Data", "DATAfile_Consolidated")
csv_files  <- list.files(csv_folder, pattern = "\\.csv$", full.names = TRUE)
trips      <- read_csv(csv_files, show_col_types = FALSE)

# ---------------------------------------------------------------------------
# Extract hour and summarise ride counts
#
# hour() from lubridate extracts the hour as an integer 0-23.
# Equivalent to DuckDB's HOUR() and SQL Server's DATEPART(HOUR, col).
# ---------------------------------------------------------------------------

hourly <- trips |>
  mutate(hour = hour(started_at)) |>
  group_by(member_casual, hour) |>
  summarise(rides = n(), .groups = "drop") |>
  group_by(member_casual) |>
  mutate(
    pct        = round(rides / sum(rides) * 100, 2),
    rider_label = recode(member_casual,
                         casual = "Casual Riders",
                         member = "Member Riders")
  )

cat("--- Peak Hours ---\n")
hourly |>
  group_by(member_casual) |>
  slice_max(rides, n = 3) |>
  arrange(member_casual, desc(rides)) |>
  select(member_casual, hour, rides, pct) |>
  print()

# ---------------------------------------------------------------------------
# Faceted area chart
#
# geom_area() fills the area under a line — more visually impactful than
# geom_line() alone for showing volume patterns.
#
# scale_x_continuous with breaks every 4 hours keeps the axis readable.
# annotate() adds text labels directly onto the plot — used here to call out
# the 8 AM and 5 PM commuter peaks on the member panel.
# ---------------------------------------------------------------------------

# Build annotation data for the commuter peaks (member panel only)
peak_annotations <- tibble(
  rider_label = "Member Riders",
  hour        = c(8, 17),
  label       = c("8 AM\ncommute", "5 PM\ncommute")
)

ggplot(hourly, aes(x = hour, y = rides, fill = member_casual)) +
  geom_area(alpha = 0.85) +
  geom_text(
    data     = peak_annotations,
    aes(x = hour, y = Inf, label = label),
    vjust    = 1.3,
    size     = 3.2,
    color    = "grey30",
    fontface = "italic",
    inherit.aes = FALSE
  ) +
  geom_vline(
    data        = peak_annotations,
    aes(xintercept = hour),
    linetype    = "dashed",
    color       = "grey50",
    linewidth   = 0.5,
    inherit.aes = FALSE
  ) +
  facet_wrap(~rider_label, ncol = 1, scales = "free_y") +
  scale_fill_manual(values = c("casual" = "#F4A261", "member" = "#2A9D8F")) +
  scale_x_continuous(
    breaks = seq(0, 23, by = 4),
    labels = c("12 AM","4 AM","8 AM","12 PM","4 PM","8 PM")
  ) +
  scale_y_continuous(labels = scales::comma) +
  labs(
    title    = "Members Ride on Schedule — Casuals Ride for Leisure",
    subtitle = "Member dual peaks at 8 AM and 5 PM reveal commuter usage;\ncasual single afternoon peak reveals recreational usage",
    x        = "Hour of Day",
    y        = "Number of Rides",
    fill     = NULL,
    caption  = "Source: Divvy Trip Data"
  ) +
  theme_minimal(base_size = 13) +
  theme(
    legend.position  = "none",
    strip.text       = element_text(face = "bold", size = 12),
    panel.spacing    = unit(1.5, "lines")
  )

ggsave(
  file.path(dirname(dirname(rstudioapi::getSourceEditorContext()$path)),
            "Visuals", "R", "09_hourly_distribution.png"),
  width = 9, height = 8, dpi = 150
)
